src/Jellyfin.LiveTv/Guide/GuideManager.cs

```diff
@@ -0,0 +1,713 @@
+        _tvDtoService = tvDtoService;
+    }
+
+    /// <inheritdoc />
+    public GuideInfo GetGuideInfo()
+    {
+        var startDate = DateTime.UtcNow;
+        var endDate = startDate.AddDays(GetGuideDays());
+
+        return new GuideInfo
+        {
+            StartDate = startDate,
+            EndDate = endDate
+        };
+    }
+
+    /// <inheritdoc />
+    public async Task RefreshGuide(IProgress<double> progress, CancellationToken cancellationToken)
+    {
+        ArgumentNullException.ThrowIfNull(progress);
+
+        await EmbyTV.EmbyTV.Current.CreateRecordingFolders().ConfigureAwait(false);
+
+        await _tunerHostManager.ScanForTunerDeviceChanges(cancellationToken).ConfigureAwait(false);
+
+        var numComplete = 0;
+        double progressPerService = _liveTvManager.Services.Count == 0
+            ? 0
+            : 1.0 / _liveTvManager.Services.Count;
+
+        var newChannelIdList = new List<Guid>();
+        var newProgramIdList = new List<Guid>();
+
+        var cleanDatabase = true;
+
+        foreach (var service in _liveTvManager.Services)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            _logger.LogDebug("Refreshing guide from {Name}", service.Name);
+
+            try
+            {
+                var innerProgress = new ActionableProgress<double>();
+                innerProgress.RegisterAction(p => progress.Report(p * progressPerService));
+
+                var idList = await RefreshChannelsInternal(service, innerProgress, cancellationToken).ConfigureAwait(false);
+
+                newChannelIdList.AddRange(idList.Item1);
+                newProgramIdList.AddRange(idList.Item2);
+            }
+            catch (OperationCanceledException)
+            {
+                throw;
+            }
+            catch (Exception ex)
+            {
+                cleanDatabase = false;
+                _logger.LogError(ex, "Error refreshing channels for service");
+            }
+
+            numComplete++;
+            double percent = numComplete;
+            percent /= _liveTvManager.Services.Count;
+
+            progress.Report(100 * percent);
+        }
+
+        if (cleanDatabase)
+        {
+            CleanDatabase(newChannelIdList.ToArray(), [BaseItemKind.LiveTvChannel], progress, cancellationToken);
+            CleanDatabase(newProgramIdList.ToArray(), [BaseItemKind.LiveTvProgram], progress, cancellationToken);
+        }
+
+        var coreService = _liveTvManager.Services.OfType<EmbyTV.EmbyTV>().FirstOrDefault();
+        if (coreService is not null)
+        {
+            await coreService.RefreshSeriesTimers(cancellationToken).ConfigureAwait(false);
+            await coreService.RefreshTimers(cancellationToken).ConfigureAwait(false);
+        }
+
+        // Load these now which will prefetch metadata
+        var dtoOptions = new DtoOptions();
+        var fields = dtoOptions.Fields.ToList();
+        dtoOptions.Fields = fields.ToArray();
```

Comment:
What's the point of this `ToList().ToArray()`?

Reply:
Right, I meat to point this out. As far as I can tell, this whole block does nothing, I just wanted to double-check before getting rid of it.

Reply:
Removed in latest commit